SIPS/sip-20.md

```diff
@@ -0,0 +1,181 @@
+        "url": "wss://example.com/text/endpoint",
+        "dataType": "text"
+      }
+    ]
+  }
+}
+```
+
+### Permission caveats
+
+Each caveat object MUST have `type` property that SHOULD be `websocket`.
+The caveat MUST have `url` and `dataType` properties.
+The `url` MUST be string.
+The `dataType` MUST be string that MUST be `"binary"` or `"text"`.
+
+- `type` property represents the type of data source.
+- `url` is the WebSocket URL.
+- `dataType` is type of data that WebSocket returns.
+
+The caveats MUST NOT have duplicate objects with the same `url` properties.
+
+The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.
+
+### Snap implementation
+
+This SIP introduces a new `onExternalData` function export that MAY be implemented by the Snap. The function is called every time the WebSocket client receives some data. The function SHOULD accept an object parameter that MUST have `data`, `type` and `source` properties.
+The parameters represent the following:
+- `data`: An object that represents the data that was received. Depending on the `data.type`, this SHOULD contain a `message` property, which SHOULD be either a `string` or an `ArrayBuffer`.
+- `type`: The type of the external data source. For this SIP, this SHOULD always be `websocket`.
+- `source`: The URL that the message was sent from. Snaps can use this to differentiate between different endpoints. This SHOULD be the exact same URL as specified in manifest.
+
+The specification of types of `onExternalData` and its parameters:
+
+```typescript
+/**
+ * A text message received from a WebSocket.
+ *
+ * @property type - The type of the message.
+ * @property message - The message as a string.
+ */
+type WebSocketTextMessage = {
+  type: 'text';
+  message: string;
+};
+
+/**
+ * A binary message received from a WebSocket.
+ *
+ * @property type - The type of the message.
+ * @property message - The message as an ArrayBuffer.
+ */
+type WebSocketBinaryMessage = {
+  type: 'binary';
+  message: ArrayBuffer;
+};
+
+/**
+ * A message received from a WebSocket.
+ *
+ * @property type - The type of the message.
+ * @property message - The message as a string or an ArrayBuffer, depending on
+ * the type.
+ */
+type WebSocketData = {
+  type: 'websocket';
+  data: WebSocketTextMessage | WebSocketBinaryMessage;
+};
+
+/**
+ * The data received from an external source.
+ *
+ * @property type - The type of the external data source.
+ * @property data - The data received from the external data source.
+ */
+type ExternalData = WebSocketData;
+
+type OnExternalDataHandlerArgs = ExternalData & {
+  source: string;
+};
+
+type OnExternalDataHandler = (args: OnExternalDataHandlerArgs) => Promise<void>;
+```
+
+#### Example
+
+```typescript
+import { OnExternalDataHandler } from '@metamask/snaps-types';
+import { assert } from '@metamask/utils';
+
+export const onExternalData: OnExternalDataHandler = ({ id, type, data, source }) => {
```

Comment:
Where does `id` come from here? It isn't defined in the above types

Reply:
Probably left from some of the previous iterations. Removed now, seems to be redundant.